test/unit/key_def.cc

```diff
@@ -457,16 +460,1330 @@ test_tuple_validate_key_parts_raw(void)
+			    bool is_unique)
+{
+	size_t p = 12 + (is_nullable ? 10 : 0) + (has_optional_parts ? 50 : 0);
+	plan(p);
+	header();
+
+	/* has_optional_parts is only valid if is_nullable. */
+	fail_unless(!has_optional_parts || is_nullable);
+
+	struct key_def *pk_def = test_key_def_new(
+		"[{%s%u%s%s}]",
+		"field", 0, "type", "unsigned");
+
+	/* Type is number to prevent using precompiled comparators. */
+	const bool last_is_nullable = has_optional_parts;
+	struct key_def *key_def = test_key_def_new(
+		"[{%s%u%s%s}{%s%u%s%s%s%b}{%s%u%s%s%s%b}]",
+		"field", 1, "type", "number",
+		"field", 2, "type", "number", "is_nullable", is_nullable,
+		"field", 3, "type", "number", "is_nullable", last_is_nullable);
+
+	struct key_def *cmp_def = key_def_merge(key_def, pk_def);
+
+	if (is_unique) {
+		/*
+		 * It's assumed that PK and SK index different parts. So we
+		 * cover cmp_def->unique_part_count < cmp_def->part_count
+		 * branch of the slowpath comparator (its last loop).
+		 */
+		fail_unless(cmp_def->unique_part_count > key_def->part_count);
+		cmp_def->unique_part_count = key_def->part_count;
+	}
+
+	/* Update has_optional_parts if the last parts can't be nil. */
+	struct key_def *keys[] = {pk_def, key_def};
+	size_t min_field_count = tuple_format_min_field_count(
+		keys, lengthof(keys), NULL, 0);
+	key_def_update_optionality(cmp_def, min_field_count);
+
+	fail_unless(cmp_def->is_nullable == is_nullable);
+	fail_unless(cmp_def->has_optional_parts == has_optional_parts);
+
+	const char *funcname = (const char *)xstrdup(tt_sprintf(
+		"tuple_compare_slowpath<%s, %s, key_def.is_unique = %s>",
+		is_nullable ? "true" : "false",
+		has_optional_parts ? "true" : "false",
+		is_unique ? "true" : "false"));
+
+	std::vector<struct tuple *> tuples_eq = {
+		/* Regular case. */
+		test_tuple_new("[%u%u%u%u]", 0, 0, 0, 0),
+		test_tuple_new("[%u%u%u%u]", 0, 0, 0, 0),
+	};
+
+	std::vector<struct tuple *> tuples_gt = {
+		/* Regular case. */
+		test_tuple_new("[%u%u%u%u]", 0, 1, 0, 0),
+		test_tuple_new("[%u%u%u%u]", 0, 0, 1, 1),
+
+		test_tuple_new("[%u%u%u%u]", 0, 1, 0, 0),
+		test_tuple_new("[%u%u%u%u]", 0, 0, 0, 1),
+
+		test_tuple_new("[%u%u%u%u]", 0, 0, 1, 0),
+		test_tuple_new("[%u%u%u%u]", 0, 0, 0, 1),
+
+		test_tuple_new("[%u%u%u%u]", 0, 0, 0, 1),
+		test_tuple_new("[%u%u%u%u]", 0, 0, 0, 0),
+	};
+
+	if (is_unique) {
+		if (is_nullable) {
+			/* Tuples are equal by SK, so PK is ignored. */
+			tuples_eq.push_back(
+				test_tuple_new("[%u%u%u%u]", 1, 0, 0, 0));
+			tuples_eq.push_back(
+				test_tuple_new("[%u%u%u%u]", 0, 0, 0, 0));
+		} else {
+			/*
+			 * FIXME: tuple_compare_slowpath has a logic I don't
+			 * quite understand. If the tuples are equal by SK and
+			 * we have no nils met, we should skip the PK comparison
+			 * and conclude the tuples are equal, but the comparator
+			 * has this `!is_nullable` condition making it compare
+			 * all parts of the key (including PK).
```

Comment:
```c
if (!is_nullable || !was_null_met)
	return 0;
```

If `NULL` was not met, we must compare SK only. But if the index is not nullable, `was_null_met` is definitely `false`, so this check of template parameter is done for the sake of optimization. I don't see any problem in this logic.

Reply:
Wow, I see the problem in the beginning of the comparator:
```c
if (is_nullable)
	end = part + key_def->unique_part_count;
else
	end = part + key_def->part_count;
```
It's really weird. Did you try to always use unique_part_count?

Reply:
Yes, it makes the vinyl MVCC fell bad in the `vinyl/iterator.test.lua` test.

Haven't digged into that yet.

Reply:
Just realized it's not a vinyl problem, but a memtx one: it has two tuples with the same value of unique second field in the output.